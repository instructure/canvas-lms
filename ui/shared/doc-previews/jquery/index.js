/*
 * Copyright (C) 2011 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import INST from 'browser-sniffer'
import I18n from 'i18n!jquery_doc_previews'
import $ from 'jquery'
import _ from 'underscore'
import htmlEscape from 'html-escape'
import '@canvas/jquery/jquery.ajaxJSON'
import {trackEvent} from '@canvas/google-analytics'
import '@canvas/jquery/jquery.instructure_misc_helpers' /*  /\$\.uniq/, capitalize */
import '@canvas/loading-image'
import sanitizeUrl from 'sanitize-url'

// first element in array is if scribd can handle it, second is if google can.
const previewableMimeTypes = {
  'application/vnd.openxmlformats-officedocument.wordprocessingml.template': [1, 1],
  'application/vnd.oasis.opendocument.spreadsheet': [1, 1],
  'application/vnd.sun.xml.writer': [1, 1],
  'application/excel': [1, 1],
  'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet': [1, 1],
  'text/rtf': [1, 1],
  'application/vnd.openxmlformats-officedocument.spreadsheetml.template': [1, 1],
  'application/vnd.sun.xml.impress': [1, 1],
  'application/vnd.sun.xml.calc': [1, 1],
  'application/vnd.ms-excel': [1, 1],
  'application/msword': [1, 1],
  'application/mspowerpoint': [1, 1],
  'application/rtf': [1, 1],
  'application/vnd.oasis.opendocument.presentation': [1, 1],
  'application/vnd.oasis.opendocument.text': [1, 1],
  'application/vnd.openxmlformats-officedocument.presentationml.template': [1, 1],
  'application/vnd.openxmlformats-officedocument.presentationml.slideshow': [1, 1],
  'text/plain': [1, 1],
  'application/vnd.openxmlformats-officedocument.presentationml.presentation': [1, 1],
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document': [1, 1],
  'application/postscript': [1, 1],
  'application/pdf': [1, 1],
  'application/vnd.ms-powerpoint': [1, 1]
}

// check to see if a file of a certan mimeType is previewable inline in the browser by either scribd or googleDocs
// ex: $.isPreviewable("application/mspowerpoint")  -> true
//     $.isPreviewable("application/rtf", 'google') -> false
$.isPreviewable = function(mimeType, service) {
  return (
    previewableMimeTypes[mimeType] &&
    (!service ||
      (!INST['disable' + $.capitalize(service) + 'Previews'] &&
        previewableMimeTypes[mimeType][{scribd: 0, google: 1}[service]]))
  )
}

$.fn.loadDocPreview = function(options) {
  return this.each(function() {
    const $this = $(this),
      opts = $.extend(
        {
          width: '100%',
          height: '400px'
        },
        $this.data(),
        options
      )

    function tellAppIViewedThisInline(serviceUsed) {
      // if I have a url to ping back to the app that I viewed this file inline, ping it.
      if (opts.attachment_view_inline_ping_url) {
        $.ajaxJSON(
          opts.attachment_view_inline_ping_url,
          'POST',
          {},
          () => {},
          () => {}
        )
        trackEvent(
          'Doc Previews',
          serviceUsed,
          JSON.stringify(opts, [
            'attachment_id',
            'submission_id',
            'mimetype',
            'crocodoc_session_url',
            'canvadoc_session_url'
          ])
        )
      }
    }

    if (opts.crocodoc_session_url) {
      const sanitizedUrl = sanitizeUrl(opts.crocodoc_session_url)
      var iframe = $('<iframe/>', {
        src: sanitizedUrl,
        width: opts.width,
        height: opts.height,
        allowfullscreen: '1',
        id: opts.id
      })
      iframe.appendTo($this)
      iframe.load(() => {
        tellAppIViewedThisInline('crocodoc')
        if ($.isFunction(opts.ready)) opts.ready()
      })
    } else if (opts.canvadoc_session_url) {
      const canvadocWrapper = $(
        '<div style="overflow: auto; resize: vertical;\
        border: 1px solid transparent; height: 100%;"/>'
      )
      canvadocWrapper.appendTo($this)

      const minHeight = opts.iframe_min_height !== undefined ? opts.iframe_min_height : '800px'
      const sanitizedUrl = sanitizeUrl(opts.canvadoc_session_url)
      var iframe = $('<iframe/>', {
        src: sanitizedUrl,
        width: opts.width,
        allowfullscreen: '1',
        css: {border: 0, overflow: 'auto', height: '99%', 'min-height': minHeight},
        id: opts.id
      })
      iframe.appendTo(canvadocWrapper)
      iframe.load(() => {
        tellAppIViewedThisInline('canvadocs')
        if ($.isFunction(opts.ready)) opts.ready()
      })
    } else if (
      (!INST.disableGooglePreviews &&
        (!opts.mimetype || $.isPreviewable(opts.mimetype, 'google')) &&
        opts.attachment_id) ||
      opts.public_url
    ) {
      // else if it's something google docs preview can handle and we can get a public url to this document.
      const loadGooglePreview = function() {
        // this handles both ssl and plain http.
        const googleDocPreviewUrl =
          '//docs.google.com/viewer?' +
          $.param({
            embedded: true,
            url: opts.public_url
          })
        if (!opts.ajax_valid || opts.ajax_valid()) {
          $(
            '<iframe src="' +
              htmlEscape(googleDocPreviewUrl) +
              '" height="' +
              opts.height +
              '" width="100%" />'
          )
            .appendTo($this)
            .load(() => {
              tellAppIViewedThisInline('google')
              if ($.isFunction(opts.ready)) {
                opts.ready()
              }
            })
        }
      }
      if (opts.public_url) {
        loadGooglePreview()
      } else if (opts.attachment_id) {
        let url = `/api/v1/files/${opts.attachment_id}/public_url.json`
        if (opts.submission_id) {
          url += '?' + $.param({submission_id: opts.submission_id})
        }
        if (opts.verifier) {
          url += `${opts.submission_id ? '&' : '?'}verifier=${opts.verifier}`
        } else {
          const match = window.location.search.match(/verifier=([^&]+)(?:&|$)/)
          const ver = match && match[1]
          if (ver) {
            url += `${opts.submission_id ? '&' : '?'}verifier=${ver}`
          }
        }
        $this.loadingImage()
        $.ajaxJSON(url, 'GET', {}, data => {
          $this.loadingImage('remove')
          if (data && data.public_url) {
            $.extend(opts, data)
            loadGooglePreview()
          }
        })
      }
    } else {
      // else fall back with a message that the document can't be viewed inline
      if (opts.attachment_preview_processing) {
        $this.html(
          '<p>' +
            htmlEscape(
              I18n.t(
                'errors.document_preview_processing',
                'The document preview is currently being processed. Please try again later.'
              )
            ) +
            '</p>'
        )
      } else {
        $this.html(
          '<p>' +
            htmlEscape(
              I18n.t(
                'errors.cannot_view_document_in_canvas',
                'This document cannot be displayed within Canvas.'
              )
            ) +
            '</p>'
        )
      }
    }
  })
}
