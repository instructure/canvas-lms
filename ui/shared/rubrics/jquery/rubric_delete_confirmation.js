/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import I18n from 'i18n!rubric_delete_confirmation'

const confirmationMessage = () => {
  if (ENV.context_asset_string.includes('course')) {
    return I18n.t(
      'Are you sure you want to delete this rubric? This action will remove this rubric association from all assignments in the current course, and delete any existing associated assessments.'
    )
  } else {
    return I18n.t(
      'Are you sure you want to delete this rubric? Any course currently associated with this rubric will still have access to it, but no new courses will be able to use it.'
    )
  }
}

export default confirmationMessage
