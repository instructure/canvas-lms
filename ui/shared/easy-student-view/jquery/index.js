/*
 * Copyright (C) 2020 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import $ from 'jquery'

const selector = '#easy_student_view'

const easy_student_view = {
  show() {
    $(selector).show()
  },

  hide() {
    $(selector).hide()
  }
}

export {easy_student_view as default}
