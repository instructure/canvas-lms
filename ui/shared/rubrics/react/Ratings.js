/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import _ from 'lodash'
import $ from 'jquery'
import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import {Text} from '@instructure/ui-text'
import {Flex} from '@instructure/ui-flex'
import I18n from 'i18n!edit_rubricRatings'

import {ratingShape, tierShape} from './types'

const pointString = (points, endOfRangePoints) => {
  if (endOfRangePoints !== null) {
    return I18n.t('%{points} to >%{endOfRangePoints} pts', {
      points: I18n.toNumber(points, {precision: 2, strip_insignificant_zeros: true}),
      endOfRangePoints: I18n.toNumber(endOfRangePoints, {
        precision: 2,
        strip_insignificant_zeros: true
      })
    })
  } else {
    return I18n.t('%{points} pts', {
      points: I18n.toNumber(points, {precision: 2, strip_insignificant_zeros: true})
    })
  }
}

export const Rating = props => {
  const {
    assessing,
    classes,
    description,
    endOfRangePoints,
    footer,
    long_description,
    points,
    onClick,
    shaderClass,
    tierColor,
    hidePoints,
    isSummary,
    selected
  } = props

  const shaderStyle = {backgroundColor: tierColor}
  const triangleStyle = {borderBottomColor: tierColor}
  const shaderClasses = classNames('shader', shaderClass)

  const ratingPoints = () => (
    <div className="rating-points">
      <Text size="small" weight="bold">
        {pointString(points, endOfRangePoints)}
      </Text>
    </div>
  )

  return (
    // eslint is unhappy here because it's not smart enough to understand that
    // when this is interact-able (via tabIndex), it will always have a role
    // eslint-disable-next-line jsx-a11y/no-static-element-interactions
    <div
      className={classes}
      onClick={assessing ? onClick : null}
      onKeyPress={e => (e.key === 'Enter' ? onClick() : null)}
      role={assessing ? 'button' : null}
      // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
      tabIndex={assessing ? 0 : null}
    >
      {hidePoints ? null : ratingPoints()}
      <div className="rating-description">
        <Text size="small" lineHeight="condensed" weight="bold">
          {description}
        </Text>
      </div>
      <Text size="small" lineHeight="condensed">
        {long_description}
      </Text>
      <div className="rating-footer">{footer}</div>
      <div
        className={shaderClasses}
        style={shaderStyle}
        aria-label={isSummary || !selected ? null : I18n.t('This rating is selected')}
      >
        <div className="triangle" style={triangleStyle} />
      </div>
    </div>
  )
}

const getCustomColor = (points, pointsPossible, customRatings) => {
  const sortedRatings = _.sortBy(customRatings, 'points').reverse()
  const scaledPoints =
    pointsPossible > 0 ? points * (sortedRatings[0].points / pointsPossible) : points
  const selectedRating = _.find(sortedRatings, rating => scaledPoints >= rating.points)
  if (selectedRating) {
    return `#${selectedRating.color}`
  } else {
    return `#${_.last(sortedRatings).color}`
  }
}

Rating.propTypes = {
  ...tierShape,
  assessing: PropTypes.bool.isRequired,
  footer: PropTypes.node,
  selected: PropTypes.bool,
  hidePoints: PropTypes.bool,
  isSummary: PropTypes.bool.isRequired,
  shaderClass: PropTypes.string
}
Rating.defaultProps = {
  footer: null,
  selected: false,
  endOfRangePoints: null, // eslint-disable-line react/default-props-match-prop-types
  hidePoints: false,
  shaderClass: null
}

const Ratings = props => {
  const {
    assessing,
    selectedRatingId,
    customRatings,
    defaultMasteryThreshold,
    footer,
    tiers,
    points,
    pointsPossible,
    hidePoints,
    onPointChange,
    isSummary,
    useRange
  } = props

  const pairs = tiers.map((tier, index) => {
    const next = tiers[index + 1]
    return {current: tier.points, next: next ? next.points : null}
  })

  const currentIndex = () => {
    if (selectedRatingId) {
      return _.findIndex(
        tiers,
        tier => tier.id === selectedRatingId && (useRange || tier.points === points)
      )
    } else {
      return pairs.findIndex(({current, next}) => {
        const currentMatch = points === current
        const withinRange = points > next && points <= current
        const zeroAndInLastRange = points === 0 && next === null
        return currentMatch || (useRange && (withinRange || zeroAndInLastRange))
      })
    }
  }

  const getRangePoints = (currentPoints, nextTier) => {
    if (nextTier) {
      return currentPoints === nextTier.points ? null : nextTier.points
    } else if (currentPoints !== 0) {
      return 0
    }
    return null
  }

  const getTierColor = selected => {
    if (!selected) {
      return 'transparent'
    }
    if (customRatings && customRatings.length > 0) {
      return getCustomColor(points, pointsPossible, customRatings)
    } else {
      return null
    }
  }

  const getShaderClass = selected => {
    if (!selected) {
      return null
    }
    if (customRatings && customRatings.length > 0) {
      return null
    }
    if (points >= defaultMasteryThreshold * 1.5) {
      return 'exceedsMasteryShader'
    } else if (points >= defaultMasteryThreshold) {
      return 'meetsMasteryShader'
    } else if (points >= defaultMasteryThreshold / 2) {
      return 'nearMasteryShader'
    } else {
      return 'wellBelowMasteryShader'
    }
  }

  const handleClick = (tier, selected) => {
    onPointChange(tier, selected)
    $.screenReaderFlashMessage(selected ? I18n.t('Rating selected') : I18n.t('Rating unselected'))
  }

  const selectedIndex = points !== undefined ? currentIndex() : null

  const visible = tiers
    .map((tier, index) => ({
      tier,
      index,
      selected: selectedIndex === index
    }))
    .filter(({selected}) => (isSummary ? selected : true))

  const ratings = visible
    .map(({tier, index}) => {
      const selected = selectedIndex === index
      const classes = classNames({
        'rating-tier': true,
        selected,
        assessing
      })

      return (
        <Flex.Item key={`tier-${index}`} width={`${100 / visible.length}%`} align="start">
          <Rating
            key={index}
            assessing={assessing}
            classes={classes}
            endOfRangePoints={useRange ? getRangePoints(tier.points, tiers[index + 1]) : null}
            footer={isSummary ? footer : null}
            onClick={() => handleClick(tier, selected)}
            shaderClass={getShaderClass(selected)}
            tierColor={getTierColor(selected)}
            hidePoints={isSummary || hidePoints}
            isSummary={isSummary}
            selected={selected}
            {...tier}
          />
        </Flex.Item>
      )
    })
    .filter(v => v !== null)

  const defaultRating = () => (
    <Rating
      key={0}
      assessing={assessing}
      classes="rating-tier"
      description={I18n.t('No details')}
      footer={footer}
      isSummary={isSummary}
      points={0}
      hidePoints={isSummary || hidePoints}
    />
  )

  const fullFooter = () =>
    isSummary || _.isNil(footer) ? null : <div className="rating-all-footer">{footer}</div>

  return (
    <div>
      <Flex>{ratings.length > 0 || !isSummary ? ratings : defaultRating()}</Flex>
      {fullFooter()}
    </div>
  )
}
Ratings.propTypes = {
  ...ratingShape,
  assessing: PropTypes.bool.isRequired,
  footer: PropTypes.node,
  onPointChange: PropTypes.func,
  isSummary: PropTypes.bool.isRequired,
  hidePoints: PropTypes.bool
}
Ratings.defaultProps = {
  footer: null,
  hidePoints: false,
  onPointChange: () => {}
}

export default Ratings
