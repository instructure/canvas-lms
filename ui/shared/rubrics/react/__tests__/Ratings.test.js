/*
 * Copyright (C) 2018 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'
import $ from 'jquery'
import {shallow} from 'enzyme'
import sinon from 'sinon'
import Ratings, {Rating} from '../Ratings'

// This is needed for $.screenReaderFlashMessageExclusive to work.
import '@canvas/rails-flash-notifications'

describe('The Ratings component', () => {
  const props = {
    assessing: false,
    footer: null,
    tiers: [
      {id: '1', description: 'Superb', points: 10},
      {id: '2', description: 'Meh', long_description: 'More Verbosity', points: 5},
      {id: '3', description: 'Subpar', points: 1}
    ],
    defaultMasteryThreshold: 10,
    assessmentRatingId: '2',
    points: 5,
    pointsPossible: 10,
    isSummary: false,
    useRange: false
  }

  const component = mods => shallow(<Ratings {...{...props, ...mods}} />)
  it('renders the root component as expected', () => {
    expect(component()).toMatchSnapshot()
  })

  it('renders the Rating sub-components as expected when range rating enabled', () => {
    const useRange = true
    component({useRange})
      .find('Rating')
      .forEach(el => expect(el.shallow()).toMatchSnapshot())
  })

  it('properly select the first matching rating when two tiers have the same point value and no ID is passed', () => {
    const tiers = [
      {description: 'Superb', points: 10},
      {description: 'Meh', points: 5},
      {description: 'Meh 2, The Sequel', points: 5},
      {description: 'Subpar', points: 1}
    ]
    const assessmentRatingId = null
    const selected = component({tiers, assessmentRatingId})
      .find('Rating')
      .map(el => el.prop('selected'))
    expect(selected).toEqual([false, true, false, false])
  })

  it('highlights the right rating when no assessmentRatingId present', () => {
    const ratings = (points, useRange = false, assessmentRatingId = null) =>
      component({points, useRange, assessmentRatingId})
        .find('Rating')
        .map(el => el.shallow().hasClass('selected'))

    expect(ratings(10)).toEqual([true, false, false])
    expect(ratings(8)).toEqual([false, false, false])
    expect(ratings(8, true)).toEqual([true, false, false])
    expect(ratings(5)).toEqual([false, true, false])
    expect(ratings(3)).toEqual([false, false, false])
    expect(ratings(3, true)).toEqual([false, true, false])
    expect(ratings(1)).toEqual([false, false, true])
    expect(ratings(0, true)).toEqual([false, false, true])
    expect(ratings(undefined)).toEqual([false, false, false])
  })

  it('calls onPointChange and flashes VO message when a rating is clicked', () => {
    const onPointChange = sinon.spy()
    const flashMock = jest.spyOn($, 'screenReaderFlashMessage')
    const el = component({onPointChange})

    el.find('Rating')
      .first()
      .prop('onClick')
      .call()
    expect(onPointChange.args[0]).toEqual([{id: '1', description: 'Superb', points: 10}, false])
    expect(flashMock).toHaveBeenCalledTimes(1)
    flashMock.mockRestore()
  })

  it('uses the right default mastery level colors', () => {
    const mastery = (points, assessmentRatingId) =>
      component({points, assessmentRatingId})
        .find('Rating')
        .map(el => el.prop('tierColor'))
    expect(mastery(10, '1')).toEqual([null, 'transparent', 'transparent'])
    expect(mastery(5, '2')).toEqual(['transparent', null, 'transparent'])
    expect(mastery(1, '3')).toEqual(['transparent', 'transparent', null])
    const shaderClasses = (points, assessmentRatingId) =>
      component({points, assessmentRatingId})
        .find('Rating')
        .map(el => el.prop('shaderClass'))
    expect(shaderClasses(10, '1')).toEqual(['meetsMasteryShader', null, null])
    expect(shaderClasses(5, '2')).toEqual([null, 'nearMasteryShader', null])
    expect(shaderClasses(1, '3')).toEqual([null, null, 'wellBelowMasteryShader'])
  })

  it('uses the right custom rating colors', () => {
    const customRatings = [
      {points: 10, color: '09BCD3'},
      {points: 5, color: '65499D'},
      {points: 1, color: 'F8971C'}
    ]
    const ratings = (points, assessmentRatingId, useRange = false) =>
      component({points, assessmentRatingId, useRange, customRatings})
        .find('Rating')
        .map(el => el.prop('tierColor'))
    expect(ratings(10, '1')).toEqual(['#09BCD3', 'transparent', 'transparent'])
    expect(ratings(5, '2')).toEqual(['transparent', '#65499D', 'transparent'])
    expect(ratings(1, '3')).toEqual(['transparent', 'transparent', '#F8971C'])
    expect(ratings(0, '3', true)).toEqual(['transparent', 'transparent', '#F8971C'])
  })

  describe('custom ratings', () => {
    const customRatings = [
      {points: 100, color: '100100'},
      {points: 60, color: '606060'},
      {points: 10, color: '101010'},
      {points: 1, color: '111111'}
    ]
    const ratings = (points, assessmentRatingId, pointsPossible = 10) =>
      component({points, assessmentRatingId, pointsPossible, customRatings, useRange: true})
        .find('Rating')
        .map(el => el.prop('tierColor'))

    it('scales points to custom ratings', () => {
      expect(ratings(10, '1')).toEqual(['#100100', 'transparent', 'transparent'])
      expect(ratings(6, '1')).toEqual(['#606060', 'transparent', 'transparent'])
      expect(ratings(5, '2')).toEqual(['transparent', '#101010', 'transparent'])
      expect(ratings(4.4, '2')).toEqual(['transparent', '#101010', 'transparent'])
      expect(ratings(1, '3')).toEqual(['transparent', 'transparent', '#101010'])
      expect(ratings(0.1, '3')).toEqual(['transparent', 'transparent', '#111111'])
      expect(ratings(0, '3')).toEqual(['transparent', 'transparent', '#111111'])
    })

    it('does not scale points if pointsPossible is 0', () => {
      expect(ratings(10, '1', 0)).toEqual(['#101010', 'transparent', 'transparent'])
      expect(ratings(4, '2', 0)).toEqual(['transparent', '#111111', 'transparent'])
    })
  })

  const ratingComponent = overrides => (
    <Rating {...props.tiers[0]} isSummary={false} assessing {...overrides} />
  )

  it('is navigable and clickable when assessing', () => {
    const onClick = sinon.spy()
    const wrapper = shallow(ratingComponent({onClick}))
    const div = wrapper.find('div').at(0)
    expect(div.prop('tabIndex')).toEqual(0)
    div.simulate('click')
    expect(onClick.called).toBe(true)
  })

  it('is not navigable or clickable when not assessing', () => {
    const onClick = sinon.spy()
    const wrapper = shallow(ratingComponent({assessing: false, onClick}))
    const div = wrapper.find('div').at(0)
    expect(div.prop('tabIndex')).toBeNull()
    expect(div.prop('role')).toBeNull()
    div.simulate('click')
    expect(onClick.called).toBe(false)
  })

  it('only renders the single selected Rating with a footer in summary mode', () => {
    const el = component({points: 5, isSummary: true, footer: <div>ow my foot</div>})
    const ratings = el.find('Rating')

    expect(ratings).toHaveLength(1)

    const rating = ratings.at(0)
    expect(rating.shallow()).toMatchSnapshot()
  })

  it('renders a default rating if none of the ratings are selected', () => {
    const el = component({
      points: 6,
      assessmentRatingId: null,
      isSummary: true,
      footer: <div>ow my foot</div>
    })
    const ratings = el.find('Rating')

    expect(ratings).toHaveLength(1)

    const rating = ratings.at(0)
    expect(rating.shallow()).toMatchSnapshot()
  })

  it('hides points on the default rating if points are hidden', () => {
    const el = component({points: 6, isSummary: true, footer: <div>ow my foot</div>})
    const rating = el.find('Rating')
    expect(rating.prop('hidePoints')).toBe(true)
  })
})
