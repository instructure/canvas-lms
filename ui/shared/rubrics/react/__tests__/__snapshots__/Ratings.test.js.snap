// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Ratings component only renders the single selected Rating with a footer in summary mode 1`] = `
<div
  className="rating-tier selected"
  onClick={null}
  onKeyPress={[Function]}
  role={null}
  tabIndex={null}
>
  <div
    className="rating-description"
  >
    <Text
      as="span"
      letterSpacing="normal"
      lineHeight="condensed"
      size="small"
      weight="bold"
      wrap="normal"
    >
      Meh
    </Text>
  </div>
  <Text
    as="span"
    letterSpacing="normal"
    lineHeight="condensed"
    size="small"
    wrap="normal"
  >
    More Verbosity
  </Text>
  <div
    className="rating-footer"
  >
    <div>
      ow my foot
    </div>
  </div>
  <div
    aria-label={null}
    className="shader nearMasteryShader"
    style={
      Object {
        "backgroundColor": null,
      }
    }
  >
    <div
      className="triangle"
      style={
        Object {
          "borderBottomColor": null,
        }
      }
    />
  </div>
</div>
`;

exports[`The Ratings component renders a default rating if none of the ratings are selected 1`] = `
<div
  className="rating-tier"
  onClick={null}
  onKeyPress={[Function]}
  role={null}
  tabIndex={null}
>
  <div
    className="rating-description"
  >
    <Text
      as="span"
      letterSpacing="normal"
      lineHeight="condensed"
      size="small"
      weight="bold"
      wrap="normal"
    >
      No details
    </Text>
  </div>
  <Text
    as="span"
    letterSpacing="normal"
    lineHeight="condensed"
    size="small"
    wrap="normal"
  />
  <div
    className="rating-footer"
  >
    <div>
      ow my foot
    </div>
  </div>
  <div
    aria-label={null}
    className="shader"
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  >
    <div
      className="triangle"
      style={
        Object {
          "borderBottomColor": undefined,
        }
      }
    />
  </div>
</div>
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 1`] = `
<div
  className="rating-tier"
  onClick={null}
  onKeyPress={[Function]}
  role={null}
  tabIndex={null}
>
  <div
    className="rating-points"
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="small"
      weight="bold"
      wrap="normal"
    >
      10 to &gt;5 pts
    </Text>
  </div>
  <div
    className="rating-description"
  >
    <Text
      as="span"
      letterSpacing="normal"
      lineHeight="condensed"
      size="small"
      weight="bold"
      wrap="normal"
    >
      Superb
    </Text>
  </div>
  <Text
    as="span"
    letterSpacing="normal"
    lineHeight="condensed"
    size="small"
    wrap="normal"
  />
  <div
    className="rating-footer"
  />
  <div
    aria-label={null}
    className="shader"
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  >
    <div
      className="triangle"
      style={
        Object {
          "borderBottomColor": "transparent",
        }
      }
    />
  </div>
</div>
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 2`] = `
<div
  className="rating-tier selected"
  onClick={null}
  onKeyPress={[Function]}
  role={null}
  tabIndex={null}
>
  <div
    className="rating-points"
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="small"
      weight="bold"
      wrap="normal"
    >
      5 to &gt;1 pts
    </Text>
  </div>
  <div
    className="rating-description"
  >
    <Text
      as="span"
      letterSpacing="normal"
      lineHeight="condensed"
      size="small"
      weight="bold"
      wrap="normal"
    >
      Meh
    </Text>
  </div>
  <Text
    as="span"
    letterSpacing="normal"
    lineHeight="condensed"
    size="small"
    wrap="normal"
  >
    More Verbosity
  </Text>
  <div
    className="rating-footer"
  />
  <div
    aria-label="This rating is selected"
    className="shader nearMasteryShader"
    style={
      Object {
        "backgroundColor": null,
      }
    }
  >
    <div
      className="triangle"
      style={
        Object {
          "borderBottomColor": null,
        }
      }
    />
  </div>
</div>
`;

exports[`The Ratings component renders the Rating sub-components as expected when range rating enabled 3`] = `
<div
  className="rating-tier"
  onClick={null}
  onKeyPress={[Function]}
  role={null}
  tabIndex={null}
>
  <div
    className="rating-points"
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="small"
      weight="bold"
      wrap="normal"
    >
      1 to &gt;0 pts
    </Text>
  </div>
  <div
    className="rating-description"
  >
    <Text
      as="span"
      letterSpacing="normal"
      lineHeight="condensed"
      size="small"
      weight="bold"
      wrap="normal"
    >
      Subpar
    </Text>
  </div>
  <Text
    as="span"
    letterSpacing="normal"
    lineHeight="condensed"
    size="small"
    wrap="normal"
  />
  <div
    className="rating-footer"
  />
  <div
    aria-label={null}
    className="shader"
    style={
      Object {
        "backgroundColor": "transparent",
      }
    }
  >
    <div
      className="triangle"
      style={
        Object {
          "borderBottomColor": "transparent",
        }
      }
    />
  </div>
</div>
`;

exports[`The Ratings component renders the root component as expected 1`] = `
<div>
  <Flex
    as="span"
    direction="row"
    display="flex"
    elementRef={[Function]}
    justifyItems="start"
    withVisualDebug={false}
    wrap="no-wrap"
  >
    <Item
      align="start"
      as="span"
      elementRef={[Function]}
      key="tier-0"
      shouldGrow={false}
      shouldShrink={false}
      width="33.333333333333336%"
    >
      <Rating
        assessing={false}
        classes="rating-tier"
        description="Superb"
        endOfRangePoints={null}
        footer={null}
        hidePoints={false}
        id="1"
        isSummary={false}
        key="0"
        onClick={[Function]}
        points={10}
        selected={false}
        shaderClass={null}
        tierColor="transparent"
      />
    </Item>
    <Item
      align="start"
      as="span"
      elementRef={[Function]}
      key="tier-1"
      shouldGrow={false}
      shouldShrink={false}
      width="33.333333333333336%"
    >
      <Rating
        assessing={false}
        classes="rating-tier selected"
        description="Meh"
        endOfRangePoints={null}
        footer={null}
        hidePoints={false}
        id="2"
        isSummary={false}
        key="1"
        long_description="More Verbosity"
        onClick={[Function]}
        points={5}
        selected={true}
        shaderClass="nearMasteryShader"
        tierColor={null}
      />
    </Item>
    <Item
      align="start"
      as="span"
      elementRef={[Function]}
      key="tier-2"
      shouldGrow={false}
      shouldShrink={false}
      width="33.333333333333336%"
    >
      <Rating
        assessing={false}
        classes="rating-tier"
        description="Subpar"
        endOfRangePoints={null}
        footer={null}
        hidePoints={false}
        id="3"
        isSummary={false}
        key="2"
        onClick={[Function]}
        points={1}
        selected={false}
        shaderClass={null}
        tierColor="transparent"
      />
    </Item>
  </Flex>
</div>
`;
